### DISCUSSION TITLE
Dual supporting cookiecutter ideas?

### DISCUSSION THREAD

**Discussion by henryiii** [copier-org/copier#1195] [Q&A] [answered]
I've got a cookiecutter template and I was wondering if it was possible to dual support cookiecutter and copier for a bit by providing both `cookiecutter.json` and `copier.yml`. I've removed the custom post- and pre-process hooks, so that's good. I can tell copier to use no `.jinja` suffix, so that's good. I think I'll need to tell copier to use the `{{...}}` file as the root. Now I'm faced with the issue of namespaces. Cookiecutter uses `cookiecutter.<option>`, while copier uses `<option>`. I didn't see a `conteext-namespace: ...` or similar option in copier, which would be the easiest way to have done this. I'm thinking I could manually build the namespace or use an Jinja extension. Thoughts? Any ideas or experience with this?

**Comment by sisp** [comment]
TBH, I'd simply migrate to Copier. Cookiecutter's activity has been essentially zero for the last year or so while Copier's has been quite high with new co-maintainers since last weekend. Also, Cookiecutter lacks many great features compared to Copier such as updating and migrations, question help texts, answer validation, Jinja includes/imports and much more. I've used Cookiecutter for years but was always unhappy because of the lack of features and low activity.

I think dual support would add quite a bit of complexity that needs to be maintained. Instead, I'd rather consider having a migration tool to automate the conversion from Cookiecutter to Copier as much as possible.

**Reply by henryiii** [reply]
Cookiecutter just got new co-maintainers too, from what I understand. I don't expect huge changes, but it's not dead. Anyway, the biggest issue for me is the name I chose for my template - `cookie`. :facepalm:. I'm trying to see if we can get it renamed, since it just moved organizations anyway. And I'd like the ability to try both out for a while before jumping all in. To me it looks possible, as long as I can solve the namespace issue.

FYI, question help texts & answer validation are what are drawing me to Copier, but I can try those out while dual supporting for a bit. Migrations, not so much, since I prefer a checking tool (I wrote `repo-review` for this) over automatic updates, but I might be swayed if it works well.

I've got full nox testing for template generation, so dual supporting doesn't bother me too much, as long as I don't have to do it forever. Once trying copier for a while, I can then move over fully and strip the namespaces.

**Reply by henryiii** [reply]
PS: A migration tool likely would be very easy to write, due to the lack of features of cookiecutter. :)

**Comment by pawamoy** [comment, marked as answer]
I think you can use https://github.com/copier-org/copier-templates-extensions to put all answers into a namespace yes.

```python
# extensions.py
from copier_templates_extensions import ContextHook


class CookiecutterNamespace(ContextHook):
    def hook(self, context):
        return {"cookiecutter": context}
```

```yaml
# copier.yml
_jinja_extensions:
- copier_templates_extensions.TemplateExtensionLoader
- extensions.py:CookiecutterNamespace
```

Remember that your users will need to install copier-templates-extensions alongside Copier, for example with pipx:

```bash
pipx inject copier copier-templates-extensions
```

Also, the context within copier.yml won't be affected I think, only in files/dirs names and file contents.

**Reply by henryiii** [reply]
Not super happy with injecting (I love `pipx run <stuff>`) but a perfectly valid thing to do while trying this out - I'm not expecting to dual support forever. I'll try it, thanks!

**Reply by pawamoy** [reply]
We could consider adding copier-templates-extensions in a Copier extra :slightly_smiling_face:

**Reply by henryiii** [reply]
Extras work well with `pipx run` since https://github.com/pypa/pipx/pull/783 ;) .

**Comment by henryiii** [comment]
Working pretty well, though I realize I don't know how to make a hidden question - basically add a variable based on other variables that doesn't show up in the questions. In cookiecutter, though it's not well documented, the convention is double underscore values are not asked of the user.

I tried this:

```yaml
year:
  type: int
  default: "{% now 'utc', '%Y' %}"
  when: false
```

But it seems it's not being added to the context. (I also tried `__year`). I'd really like to do this via static config & jinja, rather than a custom extension, if possible.

PS: Is there a way to get the current year without having to inject `jinja2_time`?

**Reply by pawamoy** [reply]
Ha, that's the great Copier question :joy: See https://github.com/orgs/copier-org/discussions/678#discussioncomment-2822746, `when: false` is not recommended nor supported.

**Reply by sisp** [reply]
You can use [Jinja include/import](https://copier.readthedocs.io/en/v8.0.0/configuring/#importing-jinja-templates-and-macros) to have computed values. The only limitation is that those are always strings as that's the render result. See https://github.com/copier-org/copier/blob/v8.0.0/tests/test_imports.py for some inspiration how this looks.

**Reply by sisp** [reply]
> PS: Is there a way to get the current year without having to inject jinja2_time?

I don't think so.

**Reply by henryiii** [reply]
> You can use [Jinja include/import](https://copier.readthedocs.io/en/v8.0.0/configuring/#importing-jinja-templates-and-macros) to have computed values

Ahh, okay. It's bit of a shift in how they would be used - but that's fine. I'll have to keep to the current method while dual supporting, but that's also fine; it's easy to add these values to in the local plugin. Once I drop dual support, then the only thing that would need `--UNSAFE` and an injection would be the year support, which I think I might be able to do a "when" on, and only ask for it if `--UNSAFE` was not passed / `jinja2_time` was not present. That's a future issue, though. I could also just update the default every year, and deal with it being in the choices, it actually was in the cookiecutter choices until very recently.

**Reply by henryiii** [reply]
PS: by shift, I mean that `{{ name_slug }}` would become `{% include 'name-slug.jinja' %}`.

**Comment by henryiii** [comment]
Okay, I think it's working! Haven't fully tested the generation yet, but it looks reasonable at first glance.

I converted all the choices into nice copier choices.

I needed these options:

```yaml
_templates_suffix: ""

_subdirectory: "{% raw %}{{cookiecutter.project_name}}{% endraw %}"

_jinja_extensions:
  - copier_templates_extensions.TemplateExtensionLoader
  - extensions.py:CookiecutterNamespace
```

And `extension.py`:

```python
from copier_templates_extensions import ContextHook
import datetime


class CookiecutterNamespace(ContextHook):
    def hook(self, context):
        context["__year"] = str(datetime.datetime.today().year)
        context["__project_slug"] = context["project_name"].lower().replace("-", "_").replace(".", "_")
        context["__type"] = "compiled" if context["backend"] in ["pybind11", "skbuild", "mesonpy", "maturin"] else "pure"
        return {"cookiecutter": context}
```

And, of course, I have to `pipx inject copier copier-templates-extensions`. And to use it:

```bash
copier copy . test_pkg --vcs-ref=HEAD --UNSAFE
```

The injection and `--UNSAFE` are annoying, compared to `pipx run cookiecutter gh:scientific-python/cookie`, but not terrible. I think I can get rid of them if I wasn't dual-supporting and was okay to ask for the year (with a yearly updated default).

**Reply by yajo** [reply]
FWIW #1179 might make `--UNSAFE` slightly less annoying.

**Reply by henryiii** [reply]
Oh, nice, much less scary sounding. And similar to VSCode's prompts and such.

**Reply by pawamoy** [reply]
Congrats on making it work @henryiii, and thank you very much for the feedback!

**Reply by henryiii** [reply]
And to support the copier answer file:

```python
context["__answers"] = context["_copier_conf"]["answers_file"]
```


And in cookiecutter:

```json
"__answers": ""
```

And the file is named `{{cookiecutter.project_name}}/{{cookiecutter.__answers}}`.

I've noticed that with copier I'm getting warnings like this on every file:

```
warning: in the working copy of 'pyproject.toml', LF will be replaced by CRLF the next time Git touches it
```

on Windows. I didn't get those with cookiecutter.

PS: really trying to avoid writing "copycutter". I've done that twice now...

**Comment by henryiii** [comment]
A few thoughts while working on this:

* I wish there was a `questions:` field in the yaml, rather than having all questions top-level and configuration options starting with a `_` - seems like an unoptimial use of yaml :shrug:. Still a ton better than json. 🤦
* Choices (dict form) seem backwards - the "key" is the display, and the "value" is the key. Looks really odd if you put a descriptive string there.
* I wonder if the output could support markdown formatting. Rich (via markdown_it_py) supports converting markdown into the terminal.
* Placeholder + default doesn't really do anything (as mentioned in the docs), but it could - it could show show the placeholder, and allow a user to type starting from scratch, but if "enter" was pressed on the placeholder, it could take the default - similar to the `[yes]/no` type prompts in various programs. That would be a good way to not force a user to erase a default, but still be able to quickly bypass it just "enter" for each.

FYI, these should not be seen as negative, just recording my thoughts on the process. The user experience (at least once copier started correctly) is drastically better. (Though that was a low bar - supporting arrow keys is already a step up!)

Hey, you are using `plumbum`, nice! Plumbum was the first open-source project I started maintaining, back in 2015. And `plumbum.colors` was my first major OS contribution. :)

**Reply by yajo** [reply]
Thanks for plumbum. It's awesome.

**Comment by henryiii** [comment]
Final result (since git doesn't notify discussions with issue/pr mentions): https://github.com/scientific-python/cookie/pull/176

**Reply by pawamoy** [reply]
This PR could serve as an excellent basis for a Cookiecutter to Copier migration guide, thanks :slightly_smiling_face:
